src/families/ethereum/transactionData.js

```diff
@@ -0,0 +1,145 @@
+// @flow
+
+import type { EthereumOperationMode, Transaction } from "./types";
+import type { Account, SubAccount } from "../../types";
+
+import abiCoder from "web3-eth-abi";
+import { NotEnoughBalance } from "@ledgerhq/errors";
+
+import contractAbis from "./contractAbis";
+
+type EthereumOperationModeDataSerializer = (
+  a: Account,
+  s: SubAccount,
+  t: Transaction
+) => Buffer;
+
+type OperationModeHandlers = {
+  [mode: EthereumOperationMode]: EthereumOperationModeDataSerializer,
+};
+
+function generateMethodSignature(functionAbi: AbiMember) {
+  const paramList = functionAbi.inputs.map((input) => input.type);
+
+  return `${functionAbi.name}(${paramList.join(",")})`;
+}
+
+function encodeParameters(functionAbi: AbiMember, params: any) {
+  const encodedParameters = functionAbi.inputs.map((input) => {
+    const value = params[input.name];
+
+    if (value === undefined) {
+      throw new Error(
+        `missing parameter "${input.name}" of type "${input.type}"`
+      );
+    }
+    const encodedParameter = abiCoder.encodeParameter(input.type, value);
+    return Buffer.from(encodedParameter.slice(2), "hex");
+  });
+
+  return Buffer.concat(encodedParameters);
+}
+
+function serializeTx(contractAbi: any, functionName: string, params: any) {
+  const functionAbi = contractAbi.find(
+    (member) => member.type === "function" && member.name === functionName
+  );
+
+  if (!functionAbi) {
+    throw new Error(
+      `no such function ${functionName} in the provided contract ABI`
+    );
+  }
+
+  const functionSignature = generateMethodSignature(functionAbi);
+  const encodedFunctionSignature = Buffer.from(
+    abiCoder.encodeFunctionSignature(functionSignature).slice(2),
+    "hex"
+  );
+  const encodedFunctionParameters = encodeParameters(functionAbi, params);
+
+  return Buffer.concat([encodedFunctionSignature, encodedFunctionParameters]);
+}
+
+const erc20Transfer: EthereumOperationModeDataSerializer = (a, s, t) => {
+  let amount;
+
+  // determining the token amount to transfer
+  if (t.useAllAmount) {
+    amount = s.balance;
+  } else {
+    if (t.amount.gt(s.balance)) {
```

Comment:
it's quite curious this part was done here. i'll investigate. serializer shouldn't have to validate^^

Reply:
i think it's a relicat of old code i moved into this generic function, we'll have to split it out back into some other place. anyway, not blocker for now